### PR TITLE
feat: sync layer cleaned up contracts

### PR DESCRIPTION
**Pull request by kelemeno**
# What ❔

- in order to provide better migration of hyperchains, we can migrate them through the l1SharedBridge
- to do this we need to enable general bridging in the L1SharedBridge, and we also need to move the holding of native tokens to their own NativeTokenVault
- this solution will also later allow to bridge L2 native tokens, as well as other assets. 

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: l1-contracts/contracts/state-transition/chain-deps/facets/Mailbox.sol [hunk 1]
```diff
@@ -234,6 +246,57 @@ contract MailboxFacet is ZkSyncHyperchainBase, IMailbox {
...
+
+        BridgehubL2TransactionRequest memory wrappedRequest = _wrapRequest(_chainId, _request);
+        canonicalTxHash = _requestL2TransactionSender(wrappedRequest);
+    }
+
+    function requestL2TransactionOnSyncLayer(
+        uint256 _chainId,
+        BridgehubL2TransactionRequest calldata _request
+    ) external returns (bytes32 canonicalTxHash) {
+        require(IBridgehub(s.bridgehub).whitelistedSyncLayers(s.chainId), "Mailbox SL: not SL");
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/state-transition/chain-deps/facets/Mailbox.sol:267]: probably block.chainid should be checked here

File: l1-contracts/contracts/bridge/L1NativeTokenVault.sol [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    /// @dev Maps token balances for each chain to prevent unauthorized spending across hyperchains.
+    /// This serves as a security measure until hyperbridging is implemented.
+    /// NOTE: this function may be removed in the future, don't rely on it!
+    mapping(uint256 chainId => mapping(address l1Token => uint256 balance)) public chainBalance;
+
+    /// @dev A mapping tokenInfo => tokenAddress
+    mapping(bytes32 => address) public tokenAddress;
+
+    /// @notice Checks that the message sender is the bridgehub.
+    modifier onlyBridge(uint256 _chainId) {
```
Comment (vladbochok) [inline l1-contracts/contracts/bridge/L1NativeTokenVault.sol:49]: Should it have `_chainId` param?

Comment (kelemeno) [reply l1-contracts/contracts/bridge/L1NativeTokenVault.sol:49]: Thank you!

File: l1-contracts/contracts/bridge/L1NativeTokenVault.sol [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    /// @dev Initializes a contract for later use. Expected to be used in the proxy
+    /// @param _owner Address which can change L2 token implementation and upgrade the bridge
+    /// implementation. The owner is the Governor and separate from the ProxyAdmin from now on, so that the Governor can call the bridge.
+    function initialize(address _owner) external reentrancyGuardInitializer initializer {
+        require(_owner != address(0), "NTV owner 0");
+        _transferOwnership(_owner);
+    }
+
+    function registerToken(address _l1Token) external onlyOwnerOrBridge {
+        bytes32 tokenInfo = keccak256(abi.encode(address(this), uint256(uint160(_l1Token))));
```
Comment (Raid5594) [inline l1-contracts/contracts/bridge/L1NativeTokenVault.sol:77]: Could you please add the comment what this logic gives us

File: l1-contracts/contracts/bridge/L1NativeTokenVault.sol [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+import {IL1SharedBridge} from "./interfaces/IL1SharedBridge.sol";
+import {ETH_TOKEN_ADDRESS, TWO_BRIDGES_MAGIC_VALUE} from "../common/Config.sol";
+
+import {IBridgehub, L2TransactionRequestTwoBridgesInner, L2TransactionRequestDirect} from "../bridgehub/IBridgehub.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @dev Vault holding L1 native ETH and ERC20 tokens bridged into the hyperchains.
+/// @dev Designed for use with a proxy for upgradability.
+contract L1NativeTokenVault is
```
Comment (Raid5594) [inline l1-contracts/contracts/bridge/L1NativeTokenVault.sol:25]: I can see that this essentially separates from the shared bridge, but in the diagram we have a separation between native token vault & the standard scenario.

Does this mean we're only considering here the tokens native to L1?
As I understand the vault for L2 native tokens has yet to be added

Comment (kelemeno) [reply l1-contracts/contracts/bridge/L1NativeTokenVault.sol:25]: Yes, we cannot handle l2 native token yet. ( You could implement it if you want )

File: l1-contracts/contracts/bridge/L1NativeTokenVault.sol [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        bytes32 tokenInfo = keccak256(abi.encode(address(this), uint256(uint160(_l1Token))));
+        L1_SHARED_BRIDGE.setAssetAddress(bytes32(uint256(uint160(_l1Token))), address(this));
+        tokenAddress[tokenInfo] = _l1Token;
+    }
+
+    /// @notice Allows bridgehub to acquire mintValue for L1->L2 transactions.
+    /// @dev If the corresponding L2 transaction fails, refunds are issued to a refund recipient on L2.
+    function bridgeBurn(
+        uint256 _chainId,
+        uint256 _mintValue,
```
Comment (Raid5594) [inline l1-contracts/contracts/bridge/L1NativeTokenVault.sol:86]: Shall we add some logic to actually use the _mintValue?

File: l1-contracts/contracts/bridge/L1NativeTokenVault.sol [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+            return abi.encode(name, symbol, decimals); // when depositing eth to a non-eth based chain it is an ERC20
+        }
+
+        (, bytes memory data1) = _token.staticcall(abi.encodeCall(IERC20Metadata.name, ()));
+        (, bytes memory data2) = _token.staticcall(abi.encodeCall(IERC20Metadata.symbol, ()));
+        (, bytes memory data3) = _token.staticcall(abi.encodeCall(IERC20Metadata.decimals, ()));
+        return abi.encode(data1, data2, data3);
+    }
+
+    function bridgeMint(address _account, uint256 _amount) external payable override {}
```
Comment (Raid5594) [inline l1-contracts/contracts/bridge/L1NativeTokenVault.sol:141]: Shall we add finalizeWithdrawal logic here?

I can see that we moved the deposit, but withdraw is still in L1SharedBridge

File: l1-contracts/contracts/bridge/L1NativeTokenVault.sol (hunk 5 above)
Comment (kelemeno) [reply l1-contracts/contracts/bridge/L1NativeTokenVault.sol:88]: Weth bridge will use mintValue, we do not actually need to use it here I think

File: l1-contracts/contracts/bridge/L1NativeTokenVault.sol (hunk 6 above)
Comment (kelemeno) [reply l1-contracts/contracts/bridge/L1NativeTokenVault.sol:143]: Yes, I'm in the process of moving the things from L1SharedBridge, I will push

File: da-contracts/contracts/CalldataDA.sol [hunk 7]
```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.24;
+
+// solhint-disable gas-custom-errors, reason-string
+
+// TODO: maybe move it here
```
Comment (vladbochok) [inline da-contracts/contracts/CalldataDA.sol:7]:
```suggestion
```

File: da-contracts/contracts/RollupL1DAValidator.sol [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+pragma solidity 0.8.24;
+
+// solhint-disable gas-custom-errors, reason-string
+
+import {IL1DAValidator, L1DAValidatorOutput} from "./IL1DAValidator.sol";
+
+import {CalldataDA} from "./CalldataDA.sol";
+
+// TODO: maybe move it here
```
Comment (vladbochok) [inline da-contracts/contracts/RollupL1DAValidator.sol:11]:
```suggestion
```

File: l1-contracts/contracts/bridgehub/Bridgehub.sol [hunk 9]
```diff
@@ -25,8 +29,11 @@ contract Bridgehub is IBridgehub, ReentrancyGuard, Ownable2StepUpgradeable, Paus
     /// @notice the asset id of Eth
     bytes32 internal immutable ETH_TOKEN_ASSET_ID;
 
+    /// @dev The chain id of L1, this contract will be deployed on multiple layers.
+    uint256 public immutable L1_CHAIN_ID;
```
Comment (Raid5594) [inline l1-contracts/contracts/bridgehub/Bridgehub.sol:33]: Won't this break the storage layout of current bridgehub? 
ETH_TOKEN_ASSET_ID should be moved lower as well

File: l1-contracts/contracts/bridgehub/Bridgehub.sol [hunk 10]
```diff
@@ -45,17 +52,32 @@ contract Bridgehub is IBridgehub, ReentrancyGuard, Ownable2StepUpgradeable, Paus
     /// @dev used to accept the admin role
     address private pendingAdmin;
 
+    IMessageRoot public override messageRoot;
```
Comment (Raid5594) [inline l1-contracts/contracts/bridgehub/Bridgehub.sol:55]: docs missing

File: l1-contracts/contracts/bridgehub/Bridgehub.sol [hunk 11]
```diff
@@ -45,17 +52,32 @@ contract Bridgehub is IBridgehub, ReentrancyGuard, Ownable2StepUpgradeable, Paus
...
+
     /// @notice Mapping from chain id to encoding of the base token used for deposits / withdrawals
     mapping(uint256 chainId => bytes32 baseTokenAssetId) public baseTokenAssetId;
 
+    ISTMDeploymentTracker public stmDeployer;
```
Comment (Raid5594) [inline l1-contracts/contracts/bridgehub/Bridgehub.sol:60]: interface vs storage variable name is confusing, shall it be the stmDeploymentTracker then?

File: l1-contracts/contracts/bridgehub/Bridgehub.sol [hunk 12]
```diff
@@ -89,11 +118,23 @@ contract Bridgehub is IBridgehub, ReentrancyGuard, Ownable2StepUpgradeable, Paus
         emit NewAdmin(previousAdmin, currentPendingAdmin);
     }
 
-    ///// Getters
+    /// @notice To set stmDeploymetTracker, only Owner. Not done in initialize, as
+    /// the order of deployment is Bridgehub, Shared bridge, and then we call this
+    function setSTMDeployer(ISTMDeploymentTracker _stmDeployer) external onlyOwner {
```
Comment (Raid5594) [inline l1-contracts/contracts/bridgehub/Bridgehub.sol:123]: Do we expect a case where this is called separately from `setAddresses`?

File: da-contracts/contracts/ValidiumL1DAValidator.sol [hunk 13]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.24;
+
+// solhint-disable gas-custom-errors, reason-string
+
+import {IL1DAValidator, L1DAValidatorOutput} from "./IL1DAValidator.sol";
+
+contract ValidiumL1DAValidator is IL1DAValidator {
+    function checkDA(
```
Comment (Raid5594) [inline da-contracts/contracts/ValidiumL1DAValidator.sol:10]: docs missing

File: l1-contracts/contracts/bridge/L1AssetRouter.sol [hunk 14]
```diff
@@ -72,7 +74,7 @@ contract L1SharedBridge is IL1SharedBridge, ReentrancyGuard, Ownable2StepUpgrade
     IL1ERC20Bridge public override legacyBridge;
 
     /// @dev A mapping chainId => bridgeProxy. Used to store the bridge proxy's address, and to see if it has been deployed yet.
```
Comment (Raid5594) [inline l1-contracts/contracts/bridge/L1AssetRouter.sol:76]: Maybe we should explain why it's been deprecated

File: l1-contracts/contracts/bridgehub/Bridgehub.sol (hunk 9 above)
Comment (kelemeno) [reply l1-contracts/contracts/bridgehub/Bridgehub.sol:33]: Its an immutable so no

**Pull request merged**